infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java

```diff
@@ -210,23 +255,123 @@ private void addMachineVolumes(
           getVolumeSubPath(workspaceId, volumeEntry.getKey(), Names.machineName(pod, container));
       subPaths.add(subPath);
 
-      container.getVolumeMounts().add(newVolumeMount(pvcName, volumePath, subPath));
-      addVolumeIfNeeded(pvcName, pod.getSpec());
+      PodSpec podSpec = pod.getSpec();
+      Optional<io.fabric8.kubernetes.api.model.Volume> volumeOpt =
+          podSpec
+              .getVolumes()
+              .stream()
+              .filter(
+                  volume ->
+                      volume.getPersistentVolumeClaim() != null
+                          && pvcName.equals(volume.getPersistentVolumeClaim().getClaimName()))
+              .findAny();
+
+      io.fabric8.kubernetes.api.model.Volume podVolume;
+      if (volumeOpt.isPresent()) {
+        podVolume = volumeOpt.get();
+      } else {
+        podVolume = newVolume(pvcName, pvcName);
+        podSpec.getVolumes().add(podVolume);
+      }
+
+      container.getVolumeMounts().add(newVolumeMount(podVolume.getName(), volumePath, subPath));
     }
   }
 
-  private void addVolumeIfNeeded(String pvcName, PodSpec podSpec) {
-    if (podSpec.getVolumes().stream().noneMatch(volume -> volume.getName().equals(pvcName))) {
-      podSpec.getVolumes().add(newVolume(pvcName, pvcName));
+  private void replacePodsVolumesWithCommon(Map<String, PodData> pods, String commonPVCName) {
+    for (PodData pod : pods.values()) {
+      Set<String> pvcVolumes = new HashSet<>();
+
+      Iterator<io.fabric8.kubernetes.api.model.Volume> volumeIterator =
```

Comment:
Consider moving this code to a separate method

Reply:
create a separate method `reducePVCSourcedVolumes`